src/idiomatic/leveraging-the-type-system/raii/mutex.md

```diff
@@ -0,0 +1,131 @@
+# Mutex
+
+In earlier examples, RAII was used to manage concrete resources like file
+descriptors. With a `Mutex`, the resource is more abstract: exclusive access to
+a value.
+
+Rust models this using a `MutexGuard`, which ties access to a critical section
+to the lifetime of a value on the stack.
+
+```rust
+#[derive(Debug)]
+struct Mutex<T> {
+    value: std::cell::UnsafeCell<T>,
+    // [...]
+}
+
+#[derive(Debug)]
+struct MutexGuard<'a, T> {
+    value: &'a mut T,
+    // [...]
+}
+
+impl<T> Mutex<T> {
+    fn new(value: T) -> Self {
+        Self {
+            value: std::cell::UnsafeCell::new(value),
+            // [...]
+        }
+    }
+
+    fn lock(&self) -> MutexGuard<T> {
+        // [...]
+        let value = unsafe { &mut *self.value.get() };
+        MutexGuard { value }
+    }
+}
+
+impl<'a, T> std::ops::Deref for MutexGuard<'a, T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        self.value
+    }
+}
+
+impl<'a, T> std::ops::DerefMut for MutexGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.value
+    }
+}
+
+impl<'a, T> Drop for MutexGuard<'a, T> {
+    fn drop(&mut self) {
+        // [...]
+        println!("drop MutexGuard");
+    }
```

Comment:
I think this is failing to demonstrate the crucial part of `MutexGuard`: That it reports back to the `Mutex` on drop, unlocking the mutex. I think we should have a stubbed-out `unlock` method on `Mutex` and call it on drop here.